### PR TITLE
Integration test for embeded Android views touch support.

### PR DESCRIPTION
**Pull request by amirh**
*This depends on #19907, #20052, and flutter/goldens#10, devicelab configuration will be added in a following PR.*

This test verifies that the synthesized motion events that get to embedded
Android view are equal to the motion events that originally hit the FlutterView.

The test app's Android code listens to MotionEvents that get to FlutterView and
to an embedded Android view and sends them over a platform channel to the Dart
code, where the events are matched.

This is what the app looks like:

android_views test app

The blue part is the embedded Android view, because it is positioned at the top
left corner, the coordinate systems for FlutterView and for the embedded view's
virtual display has the same origin (this makes the MotionEvent comparison
easier as we don't need to translate the coordinates).

The app includes the following control buttons:
  * RECORD - Start listening for MotionEvents for 3 seconds, matched/unmatched events are
    displayed in the listview as they arrive.
  * CLEAR - Clears the events that were recorded so far.
  * SAVE - Saves the events that hit FlutterView to a file.
  * PLAY FILE - Send a list of events from a bundled asset file to FlutterView.

A recorded touch events sequence is bundled as an asset in the
assets_for_android_view package which lives in the goldens repository.

When running this test with `flutter drive` the record touch sequences is
replayed and the test asserts that the events that got to FlutterView are
equivalent to the ones that got to the embedded view.

### CODE REVIEW

File: dev/integration_tests/android_views/README.md [hunk 1]
```diff
@@ -0,0 +1,32 @@
+# Integration test for touch events on embedded Android views
```
Comment (xster) [inline dev/integration_tests/android_views/README.md]: Are you adding more generic Android view embedding tests to this or is it just for touch events? (just wondering if you're going to want to use this generic name later)

File: dev/integration_tests/android_views/android/app/src/main/java/io/flutter/integration/androidviews/MainActivity.java [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+  protected void onCreate(Bundle savedInstanceState) {
+    super.onCreate(savedInstanceState);
+    GeneratedPluginRegistrant.registerWith(this);
+    getFlutterView().getPluginRegistry()
+            .registrarFor("io.flutter.integration.android_views").platformViewRegistry()
+            .registerViewFactory("simple_view", new SimpleViewFactory(getFlutterView()));
+    mMethodChannel = new MethodChannel(this.getFlutterView(), "android_views_integration");
+    mMethodChannel.setMethodCallHandler(this);
+    mFlutterViewTouchPipe = new TouchPipe(mMethodChannel, getFlutterView());
+
```
Comment (xster) [inline dev/integration_tests/android_views/android/app/src/main/java/io/flutter/integration/androidviews/MainActivity.java]: extra line

File: dev/integration_tests/android_views/android/app/src/main/java/io/flutter/integration/androidviews/MainActivity.java [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+import java.util.HashMap;
+
+import io.flutter.app.FlutterActivity;
+import io.flutter.plugin.common.MethodCall;
+import io.flutter.plugin.common.MethodChannel;
+import io.flutter.plugins.GeneratedPluginRegistrant;
+
+public class MainActivity extends FlutterActivity implements MethodChannel.MethodCallHandler {
+
```
Comment (xster) [inline dev/integration_tests/android_views/android/app/src/main/java/io/flutter/integration/androidviews/MainActivity.java]: uber nit: be consistent with extra line at top and bottom of scopes

File: dev/integration_tests/android_views/android/app/src/main/java/io/flutter/integration/androidviews/MotionEventCodec.java [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+
+import static android.view.MotionEvent.PointerCoords;
+import static android.view.MotionEvent.PointerProperties;
+
+@TargetApi(Build.VERSION_CODES.ICE_CREAM_SANDWICH)
+public class MotionEventCodec {
+
```
Comment (xster) [inline dev/integration_tests/android_views/android/app/src/main/java/io/flutter/integration/androidviews/MotionEventCodec.java]: ditto

File: dev/integration_tests/android_views/android/app/src/main/java/io/flutter/integration/androidviews/MotionEventCodec.java [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+    public static HashMap<String, Object> encode(MotionEvent event) {
+        ArrayList<HashMap<String,Object>> pointerProperties = new ArrayList<>();
```
Comment (xster) [inline dev/integration_tests/android_views/android/app/src/main/java/io/flutter/integration/androidviews/MotionEventCodec.java]: be consistent with block indents

File: dev/integration_tests/android_views/android/app/src/main/java/io/flutter/integration/androidviews/MainActivity.java [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+      case "getStoragePermission":
+        if (permissionResult != null) {
+          result.error("error", "already waiting for permissions", null);
+          return;
+        }
+        permissionResult = result;
+        getExternalStoragePermissions();
+        return;
+      case "synthesizeEvent":
+          synthesizeEvent(methodCall, result);
```
Comment (xster) [inline dev/integration_tests/android_views/android/app/src/main/java/io/flutter/integration/androidviews/MainActivity.java]: Keep consistent spacing

File: dev/integration_tests/android_views/android/app/src/main/java/io/flutter/integration/androidviews/MotionEventCodec.java [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+        map.put("y", coords.y);
+        return map;
+    }
+
+    @SuppressWarnings("unchecked")
+    public static MotionEvent decode(HashMap<String, Object> data) {
+        List<PointerProperties> pointerProperties = new ArrayList<>();
+        List<PointerCoords> pointerCoords = new ArrayList<>();
+
+        for (HashMap<String, Object> property : (List<HashMap<String, Object>>)data.get("pointerProperties")) {
```
Comment (xster) [inline dev/integration_tests/android_views/android/app/src/main/java/io/flutter/integration/androidviews/MotionEventCodec.java]: https://google.github.io/styleguide/javaguide.html#s4.6.2-horizontal-whitespace `#5`, add a space after casting or be consistent in the file.

File: dev/integration_tests/android_views/lib/main.dart [hunk 8]
```diff
@@ -0,0 +1,297 @@
...
+
+  @override
+  Widget build(BuildContext context) {
+    return new Column(
+      children: <Widget>[
+        new SizedBox(
+            height: 300.0,
+            child: new AndroidView(
+                viewType: 'simple_view',
+                onPlatformViewCreated: onPlatformViewCreated)),
```
Comment (xster) [inline dev/integration_tests/android_views/lib/main.dart]: indents and trailing commas

File: dev/integration_tests/android_views/pubspec.yaml [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+
+dependencies:
+  flutter:
+    sdk: flutter
+  flutter_driver:
+    sdk: flutter
+  path_provider: ^0.4.1
+  collection: ^1.14.6
+  assets_for_android_views:
+    path: ../../../bin/cache/pkg/goldens/dev/integration_tests/assets_for_android_views
```
Comment (xster) [inline dev/integration_tests/android_views/pubspec.yaml]: reminder to edit this before submitting

File: dev/integration_tests/android_views/lib/motion_event_diff.dart [hunk 10]
```diff
@@ -0,0 +1,191 @@
...
+  1, // UP
+  5, // POINTER_DOWN
+  6 // POINTER_UP
+];
+
+const double kDoubleErrorMargin = 0.0001;
+
+String diffMotionEvents(
+    Map<String, dynamic> originalEvent,
+    Map<String, dynamic> synthesizedEvent) {
```
Comment (xster) [inline dev/integration_tests/android_views/lib/motion_event_diff.dart]: trailing commas

File: dev/integration_tests/android_views/lib/motion_event_diff.dart [hunk 11]
```diff
@@ -0,0 +1,191 @@
...
+    diffPointerCoords(expected, actual, i, diffBuffer);
+  }
+}
+
+void diffPointerCoords(Map<String, dynamic> expected, Map<String, dynamic> actual, int pointerIdx,
+    StringBuffer diffBuffer) {
+  diffMaps(expected, actual, diffBuffer,
+      messagePrefix: '[pointerCoord $pointerIdx] ',
+      excludeKeys: <String>[
+        'size', // Currently the framework doesn't get the size form the engine.
```
Comment (xster) [inline dev/integration_tests/android_views/lib/motion_event_diff.dart]: *from

File: dev/integration_tests/android_views/lib/motion_event_diff.dart [hunk 12]
```diff
@@ -0,0 +1,191 @@
...
+      ]);
+}
+
+void diffMaps(Map<String, dynamic> expected,
+    Map<String, dynamic> actual, StringBuffer diffBuffer,
+{List<String> excludeKeys = const <String>[],
```
Comment (xster) [inline dev/integration_tests/android_views/lib/motion_event_diff.dart]: some indent issues. one per line

File: dev/integration_tests/android_views/lib/main.dart [hunk 13]
```diff
@@ -0,0 +1,297 @@
...
+  /// The list of motion events that were passed to the embedded view.
+  List<Map<String, dynamic>> embeddedViewEvents = <Map<String, dynamic>>[];
+
+  @override
+  Widget build(BuildContext context) {
+    return new Column(
+      children: <Widget>[
+        new SizedBox(
+            height: 300.0,
+            child: new AndroidView(
```
Comment (xster) [inline dev/integration_tests/android_views/lib/main.dart]: Feel free to ignore or do in a different PR: I wonder if we can integrate a 'performance test' in here. Besides the synthesized down time and event times, can we wrap this whole thing in a transparent gesture detector, record the actual time here and put that into a list as well and also record the time on the native size, pass it back along with the motionevents and make sure the time delta doesn't get past a threshold?

**Review by xster — COMMENTED** [review]
LGTM

File: dev/integration_tests/android_views/pubspec.yaml (hunk 9 above)
Comment (amirh) [reply dev/integration_tests/android_views/pubspec.yaml]: Clarified f2f - we don't publish this to pub, we're actually fetching the goldens in the driver test and reading from this path.

File: dev/integration_tests/android_views/lib/main.dart (hunk 13 above)
Comment (amirh) [reply dev/integration_tests/android_views/lib/main.dart]: I plan to add a benchmark for the roundtrip of MotionEvent->PointerEvents->MotionEvent using this integration test.

File: dev/integration_tests/android_views/android/app/src/main/java/io/flutter/integration/androidviews/MotionEventCodec.java (hunk 5 above)
Comment (amirh) [reply dev/integration_tests/android_views/android/app/src/main/java/io/flutter/integration/androidviews/MotionEventCodec.java]: done

File: dev/integration_tests/android_views/android/app/src/main/java/io/flutter/integration/androidviews/MotionEventCodec.java (hunk 7 above)
Comment (amirh) [reply dev/integration_tests/android_views/android/app/src/main/java/io/flutter/integration/androidviews/MotionEventCodec.java]: done

File: dev/integration_tests/android_views/lib/main.dart (hunk 8 above)
Comment (amirh) [reply dev/integration_tests/android_views/lib/main.dart]: done

File: dev/integration_tests/android_views/lib/motion_event_diff.dart (hunk 10 above)
Comment (amirh) [reply dev/integration_tests/android_views/lib/motion_event_diff.dart]: done

File: dev/integration_tests/android_views/lib/motion_event_diff.dart (hunk 11 above)
Comment (amirh) [reply dev/integration_tests/android_views/lib/motion_event_diff.dart]: done

File: dev/integration_tests/android_views/lib/motion_event_diff.dart (hunk 12 above)
Comment (amirh) [reply dev/integration_tests/android_views/lib/motion_event_diff.dart]: done

**Comment by amirh** [comment]
@tvolkert I split a flutter_goldens_client packages out of flutter_goldens.
Could you take a look just at: 876f907 and 99f8a95

(I'll audit the commit history into a sequence that makes more sense before landing)

**Review by tvolkert — APPROVED** [review]
LGTM

**Comment by fkorotkov** [comment]
@Hixie I think they hit an edge case and the agent stuck in a never ending retry loop on downloading a cache. It will soon timeout and the tasks will be automatically re-run. I will work on a fix for the retry issue. Sorry about that and thanks for letting me know.

**Comment by fkorotkov** [comment]
@Hixie the infinite retry thing should be fixed now. 🙌

**Pull request merged**